pkg/controllers/reconciler.go

```diff
@@ -159,6 +174,59 @@ func (r *reconciler) createObjects(ctx context.Context,
 	return result, nil
 }
 
+func imageGetVersion(image string) string {
+	version := ""
+
+	parts := strings.SplitN(filepath.Base(image), ":", 2)
+
+	if len(parts) == 2 {
+		version = parts[1]
+	}
+
+	return version
+}
+
+func imageGetName(image string) string {
+	name := ""
+
+	parts := strings.SplitN(filepath.Base(image), ":", 2)
+
+	if len(parts) == 2 {
+		name = parts[0]
+	}
+
+	return name
+}
+
+func UpgradeToImage(image string) string {
+	out := ""
+
+	if len(UpgradeToVersion) > 0 {
+		out = imageGetName(image) + UpgradeToVersion
+	} else {
+		ver, e1 := version.ParseSemantic(imageGetVersion(image))
+		minVer, e2 := version.ParseSemantic(ImageMinVersion)
+
+		if e1 == nil && e2 == nil && ver.LessThan(minVer) {
+			out = imageGetName(image) + ImageMinVersion
+		}
+	}
+
+	return out
+}
+
+func (o *wrapOnce) Upgrade(ctx context.Context, r *reconciler, devicePlugin client.Object) {
+	log := log.FromContext(ctx)
+
+	o.once.Do(func() {
+		if UpgradePlugins && r.controller.Upgrade(ctx, devicePlugin) {
+			if err := r.Update(ctx, devicePlugin); err != nil {
+				log.Error(err, "unable to update devicePlugin")
+			}
+		}
+	})
+}
```

Comment:
If this function depends on global state already (meaning `UpgradePlugins`) then using the global `once` directly won't make the situation worse, I guess. Like in
```suggestion
func upgradeOnce(ctx context.Context, r *reconciler, devicePlugin client.Object) {
	log := log.FromContext(ctx)

	once.Do(func() {
		if UpgradePlugins && r.controller.Upgrade(ctx, devicePlugin) {
			if err := r.Update(ctx, devicePlugin); err != nil {
				log.Error(err, "unable to update devicePlugin")
			}
		}
	})
}
```

Reply:
@rojkov 
Thanks! Dropped wrapOnce.